src/JuMPArray.jl

```diff
@@ -5,67 +5,85 @@
-        idx[i] = convert(Int,(indices[i]-starts)/steps) + 1
+        inner = quote
+            idxset = indexsets[$i]
+            ret = Dict{eltype(idxset), Int}()
+        end
+        tupelem = indexsets.parameters[i]
+        if !(tupelem == UnitRange{Int} || tupelem == StepRange{Int})
+            inner = quote
+                $inner
+                cnt = 1
+                for x in idxset
+                    ret[x] = cnt
+                    cnt += 1
+                end
+                ret
+            end
+        end
+        push!(dicttuple.args, inner)
     end
-    return d.innerArray[idx...]
+    :(JuMPArray(innerArray, indexsets, $dicttuple, Dict{Symbol,Any}()))
 end
 
-function Base.setindex!{T,N}(d::JuMPArray{T,N,UnitRange{Int}},val::T,indices::Int...)
-    length(indices) == N || error("Wrong number of indices for ",d.name,", expected ",length(d.indexsets))
-    idx = Array(Int, N)
-    for i in 1:N
-        idx[i] = indices[i] - start(d.indexsets[i]) + 1
+@generated function Base.getindex{T,N,NT<:NTuple}(d::JuMPArray{T,N,NT}, idx...)
+    indexing = Any[]
+    for (i,S) in enumerate(NT.parameters)
+        if S == UnitRange{Int}
+            push!(indexing, quote
+                rng = d.indexsets[$i]
+                I = idx[$i]
+                first(rng) <= I <= last(rng) || throw(BoundsError())
+                I - (start(rng) - 1)
+            end)
+        elseif S == StepRange{Int}
+            push!(indexing, quote
+                rng = $(d.indexsets[i])
+                I = idx[$i]
+                first(rng) <= I <= last(rng) || throw(BoundsError())
+                d, r = divrem(I - start(rng), step(rng))
+                r == 0 || throw(BoundsError())
+                d + 1
+            end)
+        else
+            push!(indexing, :(d.lookup[$i][idx[$i]]::Int))
+        end
     end
-    d.innerArray[idx...] = val
+    Expr(:call, :getindex, :(d.innerArray), indexing...)
 end
 
-function Base.setindex!{T,N}(d::JuMPArray{T,N,StepRange{Int,Int}},val::T,indices::Int...)
-    length(indices) == N || error("Wrong number of indices for ",d.name,", expected ",length(d.indexsets))
-    idx = Array(Int, N)
-    steps  = 0
-    starts = 0
-    for i in 1:N
-        steps  = step(d.indexsets[i])
-        starts = start(d.indexsets[i])
-        idx[i] = convert(Int,(indices[i]-starts)/steps) + 1
+@generated function Base.setindex!{T,N,NT<:NTuple}(d::JuMPArray{T,N,NT}, v::T, idx...)
+    indexing = Any[]
+    for (i,S) in enumerate(NT.parameters)
+        if S == UnitRange{Int}
+            push!(indexing, quote
+                rng = d.indexsets[$i]
+                I = idx[$i]
+                first(rng) <= I <= last(rng) || throw(BoundsError())
+                I - (start(rng) - 1)
+            end)
+        elseif S == StepRange{Int}
+            push!(indexing, quote
+                rng = $(d.indexsets[i])
+                I = idx[$i]
+                first(rng) <= I <= last(rng) || throw(BoundsError())
```

Comment:
This ``BoundsError`` doesn't print a useful error message as it does for with Julia arrays. Would be nice to get that working.